search.go

```diff
@@ -1,175 +1,152 @@
+// It should be implemented by a version-specific executor.
+//
+// [Search]: https://openorg/docs/latest/api-reference/search/
+type Search interface {
+	Search(ctx context.Context, req *SearchRequest) (*SearchResponse, error)
+}
+
+// SearchRequest is a domain model union type for all the fields of a Search request across
+// all supported OpenSearch version.
+// Currently supported versions are:
+//   - OpenSearch 2
+//
+// This SearchRequest is intended to be used along with a version-specific executor such
+// as [opensearchtools.osv2.Executor]. For example:
 //
-//	req := NewSearchRequest()
-//	req.AddIndices("example_index")
-//	req.SetQuery(search.NewTermQuery("field", "basic")
-//	results, err := req.Do(context.Background(), client)
+//	searchReq := NewSearchRequest().
+//		WithQuery(NewTermQuery("field", "term"))
+//	searchResp, err := osv2Executor.Search(ctx, searchReq)
+//
+// An error can be returned if
+//   - The request to OpenSearch fails
+//   - The results json cannot be unmarshalled
 type SearchRequest struct {
-	Query search.Query
+	// Query to be performed by the search
+	Query Query
+
+	// Index(s) to be targeted by the search
 	Index []string
-	Size  int
-	Sort  []*search.Sort
+
+	// Size of results to be returned
+	Size int
+
+	// Sort(s) to order the results returned
+	Sort []*Sort
 }
 
 // NewSearchRequest instantiates a SearchRequest with a Size of -1.
 // Any negative value for SearchRequest.Size will be ignored and not included in the source.
 // Opensearch by default, if no size is included in a search request, will limit the results to 10 documents.
 // A NewSearchRequest will search across all indices and return the top 10 documents with the default [sorting].
 //
-// [sorting]: https://opensearch.org/docs/latest/opensearch/search/sort/
+// [sorting]: https://openorg/docs/latest/opensearch/search/sort/
 func NewSearchRequest() *SearchRequest {
 	return &SearchRequest{Size: -1}
 }
 
-// MarshalJSON marshals the SearchRequest into the JSON shape expected by OpenSearch.
-// Implemented to match the [json.Marshaler] interface.
-func (r *SearchRequest) MarshalJSON() ([]byte, error) {
-	source := make(map[string]any)
-	if r.Query != nil {
-		queryJSON, jErr := r.Query.ToOpenSearchJSON()
-		if jErr != nil {
-			return nil, jErr
-		}
-
-		source["query"] = json.RawMessage(queryJSON)
-	}
-
-	if r.Size >= 0 {
-		source["size"] = r.Size
-	}
-
-	if len(r.Sort) > 0 {
-		sorts := make([]json.RawMessage, len(r.Sort))
-		for i, s := range r.Sort {
-			sortJSON, jErr := s.ToOpenSearchJSON()
-			if jErr != nil {
-				return nil, jErr
-			}
-
-			sorts[i] = sortJSON
-		}
-
-		source["sort"] = sorts
-	}
-
-	return json.Marshal(source)
-}
-
-// AddIndices sets the index list for the request.
-func (r *SearchRequest) AddIndices(indices ...string) *SearchRequest {
+// WithIndices sets the index list for the request.
+func (r *SearchRequest) WithIndices(indices ...string) *SearchRequest {
```

Comment:
nit: `WithIndices()` has a connotation of replacing any existing values.

Maybe this one case should be `AddIndices()`?

Reply:
Guess I got over eager with the changing.